docker/Dockerfile

```diff
@@ -0,0 +1,64 @@
+ARG BUILDER_IMAGE=python:3.9-slim-bullseye
+FROM ${BUILDER_IMAGE} as builder
+
+USER root
+WORKDIR /
+
+SHELL [ "/bin/bash", "-c" ]
+
+COPY . /root/madanalysis5
+
+# Set PATH to pickup virtualenv by default
+ENV PATH=/usr/local/venv/bin:"${PATH}"
+RUN apt-get -qq -y update && \
+    apt-get -qq -y install --no-install-recommends \
+      gcc \
+      g++ \
+      make \
+      zlib1g \
+      bash-completion \
+      python3-dev \
+      less \
+      tree \
+      wget \
+      curl \
+      gnuplot \
+      git && \
+    apt-get -y clean && \
+    apt-get -y autoremove && \
+    rm -rf /var/lib/apt/lists/* && \
+    python -m venv /usr/local/venv && \
+    . /usr/local/venv/bin/activate && \
+    python -m pip --no-cache-dir install --upgrade pip setuptools wheel && \
+    python -m pip --no-cache-dir install pip-tools && \
+    python -m pip list && \
+    python -m piptools compile \
+        --generate-hashes \
+        --output-file /root/madanalysis5/requirements.lock \
+        /root/madanalysis5/docker/requirements.txt && \
+    python -m pip --no-cache-dir install --upgrade --requirement /root/madanalysis5/requirements.lock && \
```

Comment:
Using piptools here to create a hash-level lockfile (`/root/madanalysis5/requirements.lock`) so that the Python environment that is shipped is fully reproducible and explained (`piptools compile` will add comments to the generated lockfile explaining why things were added).